### PR TITLE
add cpp ut and python ut for GRPC

### PR DESCRIPTION
**Pull request by szc321**
Signed-off-by: szc321 <zongcai.shi@intel.com>

Issue Number: 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [ ] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by szc321** [comment]
the cpp ut and python ut can run normally without export the env: GRPC_ENABLE_FORK_SUPPORT=true, GRPC_POLL_STRATEGY=poll

File: src/lava/magma/runtime/message_infrastructure/test/CMakeLists.txt [hunk 1]
```diff
@@ -12,9 +12,11 @@ include_directories(../message_infrastructure/csrc)
 
 add_executable(
   test_messaging_infrastructure
-  cpptest/test_multiprocessing.cc
+  # cpptest/test_multiprocessing.cc
+  # cpptest/test_grpc_channel.cc
+  cpptest/test_grpc_delivery.cc
   # cpptest/test_channel.cc
-  cpptest/test_shm_delivery.cc
+  # cpptest/test_shm_delivery.cc
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/CMakeLists.txt:19]: don't change other tests

File: src/lava/magma/runtime/message_infrastructure/test/cpptest/test_grpc_delivery.cc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                                a2_to_a1,
+                                std::placeholders::_1);
+  auto target_fn_a2 = std::bind(&target_fn2,
+                                loop,
+                                a1_to_a2,
+                                a2_to_a1,
+                                std::placeholders::_1);
+  int actor1 = mp.BuildActor(target_fn_a1);
+  int actor2 = mp.BuildActor(target_fn_a2);
+  auto to_a1   = mp_to_a1->GetSendPort();
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/cpptest/test_grpc_delivery.cc:121]: code style

File: src/lava/magma/runtime/message_infrastructure/test/cpptest/test_grpc_delivery.cc [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  metadata->dims[0] = 1;
+  metadata->strides[0] = 1;
+  metadata->mdata =
+    reinterpret_cast<char*>
+    (malloc(sizeof(int64_t)));
+  *reinterpret_cast<int64_t*>(metadata->mdata) = 1;
+  MultiProcessing mp;
+  int loop = 10000;
+  AbstractChannelPtr mp_to_a1 = GetChannelFactory().GetRPCChannel(
+  "127.13.2.11", 8001, "mp_to_a1", "mp_to_a1", 6);
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/cpptest/test_grpc_delivery.cc:100]: code style

File: src/lava/magma/runtime/message_infrastructure/test/cpptest/test_grpc_delivery.cc [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    mptr->strides[0], mptr->strides[1], mptr->strides[2], mptr->strides[3],
+    mptr->strides[4]);
+    LAVA_DUMP(1, "mdata: %p, *mdata: %ld\n", mptr->mdata,
+              *reinterpret_cast<int64_t*>(mptr->mdata));
+    free(reinterpret_cast<char*>(metadata->mdata));
+    metadata = mptr;
+  }
+
+  const clock_t end_time = std::clock();
+  if (*reinterpret_cast<int64_t*>(metadata->mdata) != expect_result) {
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/cpptest/test_grpc_delivery.cc:149]: put this checking after `mp.strop`

File: src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        data[0] = data[0] + 1
+        to_a1.send(data)
+        loop -= 1
+
+    from_a1.join()
+    to_a1.join()
+    actor.status_stopped()
+
+
+class TestChannel(unittest.TestCase):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py:88]: rename as `TestGrpcDelivery`

File: src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        data = prepare_data()
+        expect_result = prepare_data()
+        expect_result[0] = (1 + 3 * loop)
+        loop_start_time = datetime.now()
+        while loop:
+            to_a1.send(data)
+            data = from_a1.recv()
+            loop -= 1
+            print("mainloop:", loop)
+            print("mp_recv_data_from_actor1", data)
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py:142]: remove these prints

File: src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        loop_end_time = datetime.now()
+        if not np.array_equal(expect_result, data):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py:144]: put this checking after `mp.stop`

File: src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        to_a1.send(data)
+        loop -= 1
+
+    from_a1.join()
+    to_a1.join()
+    actor.status_stopped()
+
+
+class TestChannel(unittest.TestCase):
+    def test_grpcchannel(self):
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/test/test_grpc_delivery.py:89]: add skipIf condition with SupportGPRCChannel

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 9]
```diff
@@ -117,6 +118,7 @@ void GrpcRecvPort::Join() {
     done_ = true;
     service_ptr_->Stop();
     server_->Shutdown();
+    GetGrpcManager().Release(url_);
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:121]: i think it is useless.
the child process invokes this Join func, not parent process.

Comment (szc321) [reply src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:121]: Yes, I tried it, release action doesn't help

**Pull request merged**